saleor/graphql/checkout/dataloaders.py

```diff
@@ -86,101 +90,153 @@ def with_variants_products_collections(results):
-                                        variant_listing_promotion_rule=listing_rule,
-                                        promotion=rule_id_to_promotion_map[
-                                            listing_rule.promotion_rule_id
-                                        ],
+                            )
+                            rule_map = dict(zip(rule_ids, promotion_rules))
+                            rule_id_to_promotion_map = dict(zip(rule_ids, promotions))
+                            rule_id_to_rule_translation = dict(
+                                zip(rule_ids, rule_translations)
+                            )
+                            rule_id_to_promotion_translation = dict(
+                                zip(rule_ids, promotion_translations)
+                            )
+
+                            lines_info_map = defaultdict(list)
+                            voucher_infos_map = {
+                                voucher_info.voucher.code: voucher_info
+                                for voucher_info in voucher_infos
+                                if voucher_info
+                            }
+                            for checkout, lines in zip(checkouts, checkout_lines):
+                                for line in lines:
+                                    channel_listing = channel_listings_map[
+                                        (line.variant_id, checkout.channel_id)
+                                    ]
+                                    listing_promotion_rules = (
+                                        listing_promotion_rules_map[channel_listing.id]
+                                        if channel_listing
+                                        else []
                                     )
-                                    for listing_rule in listing_promotion_rules
-                                ]
-                                lines_info_map[checkout.pk].append(
-                                    CheckoutLineInfo(
-                                        line=line,
-                                        variant=variants_map[line.variant_id],
-                                        channel_listing=channel_listing,
-                                        product=products_map[line.variant_id],
-                                        product_type=product_types_map[line.variant_id],
-                                        collections=collections_map[line.variant_id],
-                                        discounts=checkout_lines_discounts_map[line.id],
-                                        rules_info=rules_info,
-                                        tax_class=tax_class_map[line.variant_id],
-                                        channel=channel_maps[checkout.channel_id],
+                                    rules_info = [
+                                        VariantPromotionRuleInfo(
```

Comment:
How about creating a dataloader for creating PromotionRulesInfo and use inside this dataloader?